File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCheckZeroSizeCollection.vb
Imports System.Collections.Generic
Imports System.Linq

Namespace Tests.Diagnostics
    Class FooProperty
        Public ReadOnly Property Length() As Integer
            Get
                Return 0
            End Get
        End Property

        Public ReadOnly Property LongLength() As Integer
            Get
                Return 0
            End Get
        End Property

        Public ReadOnly Property Count() As Integer
            Get
                Return 0
            End Get
        End Property
    End Class

    Class FooMethod
        Public Function Count() As Integer
            Return 0
        End Function
    End Class

    Class DummyHolder
        Public Enumerable As List(Of String)
        Public Function GetEnumerable() As List(Of String)
            Return Nothing
        End Function

        Public Array As String()
        Public Function GetArray() As String()
            Return Nothing
        End Function

        Public Holder As DummyHolder
        Public Function GetHolder() As DummyHolder
            Return Nothing
        End Function
    End Class

    Class Program
        Const ConstField_Zero As Integer = 0
        Const ConstField_NonZero As Integer = 1

        Public Sub TestCountMethod()
            Const LocalConst_Zero As Integer = 0
            Const LocalConst_NonZero As Integer = 1
            Dim LocalVariable As Integer = 0
            Dim Result As Boolean

            Dim SomeEnumerable As IEnumerable(Of String) = New List(Of String)()

            Result = Enumerable.Count(SomeEnumerable) >= 0 ' Noncompliant {{The 'Count' of 'IEnumerable(Of T)' always evaluates as 'True' regardless the size.}}
            '        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            Result = SomeEnumerable.Count(Function(foo) True) >= 0 ' Noncompliant {{The 'Count' of 'IEnumerable(Of T)' always evaluates as 'True' regardless the size.}}
            Result = SomeEnumerable?.Count() >= 0 ' Noncompliant
            Result = SomeEnumerable.Count() >= 1
            Result = SomeEnumerable.Count() >= LocalVariable
            Result = SomeEnumerable.Count() >= -1 ' Noncompliant {{The 'Count' of 'IEnumerable(Of T)' always evaluates as 'True' regardless the size.}}
            Result = SomeEnumerable.Count() <= 0
            Result = SomeEnumerable.Count() < 0 ' Noncompliant {{The 'Count' of 'IEnumerable(Of T)' always evaluates as 'False' regardless the size.}}
            Result = 0 >= SomeEnumerable.Count()

            Result = SomeEnumerable.Count() >= LocalConst_Zero ' Noncompliant
            Result = SomeEnumerable.Count() >= ConstField_NonZero
            Result = SomeEnumerable.Count() >= ConstField_Zero ' Noncompliant
            Result = SomeEnumerable.Count() >= LocalConst_NonZero

            Result = (SomeEnumerable.Count()) >= (0) ' Noncompliant
            Result = ((((SomeEnumerable).Count())) >= ((0))) ' Noncompliant
            '         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            Result = 0 <= SomeEnumerable.Count() ' Noncompliant
            '        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
            Dim NonEnumerable = New FooMethod()
            Result = NonEnumerable.Count() >= 0
        End Sub

        Public Sub TestComplexAccess()
            Dim SomeArray = New String(-1) {}
            Dim SomeEnumerable = New List(Of String)()
            Dim Holder = New DummyHolder()

            Dim Result As Boolean
            Result = Holder.GetHolder().GetHolder().GetArray().Count() >= 0 ' Noncompliant
            Result = Holder.GetHolder()?.GetHolder()?.GetArray()?.Length >= 0 ' Noncompliant
            Result = Holder.GetHolder()?.GetHolder().Holder.Array.Length >= 0 ' Noncompliant
            Result = Holder.GetHolder().GetHolder().Holder.Enumerable.AsEnumerable.Count(Function(foo) True) >= 0 ' Noncompliant
            Result = (Holder.GetHolder()?.GetHolder())?.GetArray()?.Length >= 0 ' Noncompliant
        End Sub

        Public Sub TestCountProperty()
            Dim SomeCollection = New List(Of String)()
            Dim Result As Boolean = SomeCollection.Count >= 0 ' Noncompliant {{The 'Count' of 'ICollection' always evaluates as 'True' regardless the size.}}
            '                       ^^^^^^^^^^^^^^^^^^^^^^^^^
            Dim NonCollection = New FooProperty()
            Result = NonCollection.Count >= 0
        End Sub

        Public Sub TestLengthProperty()
            Dim SomeArray = New String(-1) {}
            Dim Result As Boolean

            Result = SomeArray.Length >= 0 ' Noncompliant {{The 'Length' of 'Array' always evaluates as 'True' regardless the size.}}
            '        ^^^^^^^^^^^^^^^^^^^^^
            Result = SomeArray.LongLength >= 0 ' Noncompliant {{The 'LongLength' of 'Array' always evaluates as 'True' regardless the size.}}
            '        ^^^^^^^^^^^^^^^^^^^^^^^^^
            Dim NonArray = New FooProperty()
            Result = NonArray.Length >= 0
            Result = NonArray.LongLength >= 0
        End Sub

        Public Sub TestInterfacesAndReadonlyCollections(List As IList(Of Integer), Collection As ICollection(Of Integer), ReadonlyCollection As IReadOnlyCollection(Of Integer), ReadonlyList As IReadOnlyList(Of Integer))
            Dim Result As Boolean
            Dim SortedSet As New SortedSet(Of Double)

            Result = List.Count >= 0 ' Noncompliant

            Result = Collection.Count >= 0 ' Noncompliant

            Result = ReadonlyCollection.Count >= 0 ' Noncompliant

            Result = ReadonlyList.Count >= 0 ' Noncompliant

            Result = SortedSet.Count >= 0 ' Noncompliant
        End Sub

    End Class

    Class OnString
        Shared Function LengthWithoutMeaning(str As String) As Boolean
            Return str.Length < -3 ' Noncompliant {{The 'Length' of 'String' always evaluates as 'False' regardless the size.}}
        End Function
    End Class

End Namespace

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCheckZeroSizeCollection.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Tests.Diagnostics
{
    class FooProperty
    {
        public int Length => 0;

        public int LongLength => 0;

        public int Count => 0;
    }

    class FooMethod
    {
        public int Count() => 0;
    }

    class DummyHolder
    {
        public List<string> Enumerable;
        public List<string> GetEnumerable() { return null; }

        public string[] Array;
        public string[] GetArray() { return null; }

        public DummyHolder Holder;
        public DummyHolder GetHolder() { return null; }
    }

    class Program
    {
        const double ConstField_Double_Zero = 0;
        const int ConstField_Zero = 0;
        const int ConstField_NonZero = 1;

        public void TestCountMethod()
        {
            const int localConst_Zero = 0;
            const int localConst_NonZero = 1;
            int localVariable = 0;
            bool result;

            var someEnumerable = new List<string>();

            result = someEnumerable.Count() >= 0; // Noncompliant {{The 'Count' of 'IEnumerable<T>' always evaluates as 'True' regardless the size.}}
            //       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
            result = someEnumerable.Count(foo => true) >= 0; // Noncompliant
            result = someEnumerable?.Count() >= 0; // Noncompliant

            result = someEnumerable.Count() >= 1;
            result = someEnumerable.Count() >= localVariable;
            result = someEnumerable.Count() >= -1; // Noncompliant {{The 'Count' of 'IEnumerable<T>' always evaluates as 'True' regardless the size.}}
            result = someEnumerable.Count() <= 0;
            result = someEnumerable.Count() < 0; // Noncompliant {{The 'Count' of 'IEnumerable<T>' always evaluates as 'False' regardless the size.}}
            result = 0 >= someEnumerable.Count();

            result = someEnumerable.Count() >= localConst_Zero; // Noncompliant
            result = someEnumerable.Count() >= ConstField_NonZero;
            result = someEnumerable.Count() >= ConstField_Zero; // Noncompliant
            result = someEnumerable.Count() >= localConst_NonZero;
            result = someEnumerable.Count() >= ConstField_Double_Zero; // Compliant, double is ignored, arguably FN

            result = (someEnumerable.Count()) >= (0); // Noncompliant
            result = ((((someEnumerable).Count())) >= ((0))); // Noncompliant
            //        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            result = 0 <= someEnumerable.Count(); // Noncompliant
            //       ^^^^^^^^^^^^^^^^^^^^^^^^^^^

            result = result.ToString() == "Not integer";
            result = (localVariable = 42) != -1;

            if ((localVariable = 42) != -1) { }

            var nonEnumerable = new FooMethod();
            result = nonEnumerable.Count() >= 0;
        }

        public void TestComplexAccess()
        {
            var someArray = new string[0];
            var someEnumerable = new List<string>();
            var holder = new DummyHolder();

            bool result;
            result = holder.GetHolder().GetHolder().GetArray().Count() >= 0; // Noncompliant
            result = holder.GetHolder()?.GetHolder()?.GetArray()?.Length >= 0; // Noncompliant
            result = holder.GetHolder()?.GetHolder().Holder.Array.Length >= 0; // Noncompliant
            result = holder.GetHolder()?.GetHolder().Holder.Enumerable.Count(foo => true) >= 0; // Noncompliant
            result = (holder.GetHolder()?.GetHolder())?.GetArray()?.Length >= 0; // Noncompliant
        }

        public void TestCountProperty()
        {
            var someCollection = new List<string>();
            bool result = someCollection.Count >= 0; // Noncompliant {{The 'Count' of 'ICollection' always evaluates as 'True' regardless the size.}}
            //            ^^^^^^^^^^^^^^^^^^^^^^^^^

            var nonCollection = new FooProperty();
            result = nonCollection.Count >= 0;
        }

        public void TestLengthProperty()
        {
            var someArray = new string[0];
            bool result;

            result = someArray.Length >= 0; // Noncompliant {{The 'Length' of 'Array' always evaluates as 'True' regardless the size.}}
            //       ^^^^^^^^^^^^^^^^^^^^^

            result = someArray.LongLength >= 0; // Noncompliant {{The 'LongLength' of 'Array' always evaluates as 'True' regardless the size.}}
            //       ^^^^^^^^^^^^^^^^^^^^^^^^^

            var nonArray = new FooProperty();
            result = nonArray.Length >= 0;
            result = nonArray.LongLength >= 0;
        }

        private void TestInterfacesAndReadonlyCollections(IList<int> list, ICollection<int> collection, IReadOnlyCollection<int> readonlyCollection, IReadOnlyList<int> readonlyList)
        {
            SortedSet<double> sortedSet = new SortedSet<double>();

            bool result;

            result = list.Count >= 0; // Noncompliant

            result = collection.Count >= 0; // Noncompliant

            result = readonlyCollection.Count >= 0; // Noncompliant

            result = readonlyList.Count >= 0; // Noncompliant

            result = sortedSet.Count >= 0; // Noncompliant
        }
    }

    class OnString
    {
        static bool LengthWithoutMeaning(string str)
        {
            return str.Length < -3; // Noncompliant {{The 'Length' of 'String' always evaluates as 'False' regardless the size.}}
        }
    }
}
